scripts/malleus.sh

```diff
@@ -12,14 +17,16 @@ case $1 in
         --participants) PARTICIPANTS=$optvalue;;
         --senders) SENDERS=$optvalue;;
         --audio-senders) AUDIO_SENDERS=$optvalue;;
+        --senders-per-node) SENDERS_PER_NODE=$optvalue;;
+        --receivers-per-node) RECEIVERS_PER_NODE=$optvalue;;
         --duration) DURATION=$optvalue;;
         --room-name-prefix) ROOM_NAME_PREFIX=$optvalue;;
         --hub-url) HUB_URL=$optvalue;;
         --instance-url) INSTANCE_URL=$optvalue;;
         --regions) REGIONS=$optvalue;;
+        --skip-mutation) SKIP_MUTATION=1;;
```

Comment:
What does `mutation` refer to?

Reply:
Even after seeing the file below, I'm not sure it's clear

Reply:
After re-reading it more I get it's about changing an existing mode, but I'm not sure what it implies to skip or not skip mutation.

Reply:
It means to change the settings of the Selenium node, to configure it to have N browsers, set up to be senders or receivers.  Any suggestion for a better name?

Reply:
Is this to say: selenium nodes have some default setting and this would cause the script to change the state of a node? Is the state permanent? Is it required if, say, `senders-per-node` is set to a number that doesn't match what the current node is set up for?

Reply:
Yes, Selenium nodes (at least in our Grid setup) by default run only a single browser, and don't have the `applicationName` field set in their configuration.

The script changes the nodes to run either `senders-per-node` or `receivers-per-node` browsers, distinguishing them with `applicationName`.

The changes are permanent, currently, yes.  Do you think I should have the script revert them when it's done, instead?

Reply:
Ok, I'm still confused by the semantics of `skip-mutation`, how does it work to skip mutation if the number of browsers on the node don't match what was requested?

Reply:
Well, as long as there are sufficiently many nodes marked with the appropriate `applicationName` values, things will work.  That said, if they *don't* have `applicationName` values, it'll just fail.

My goal here was to let things still (kind of) work for people who might be using Selenium grids configured differently than ours (since this is in the open-source project), but perhaps this needs a little more thought?

Reply:
Yeah I guess I'm missing context here--I'm not aware of all the semantics around `applicationName` either.  Are there applications we're talking about here other than selenium?

Reply:
`applicationName` is just a (poorly-documented) way for a Selenium node to indicate an endpoint property more specific than the browser type.  When a client requests a browser, it can (optionally) indicate that it needs a browser that has the specified `applicationName` value configured for it.  (There's no semantics from Selenium itself, it's only a way of binding node configuration to client requests.)

This script is using it to distinguish "sender" nodes (i.e. ones that will launch at most `senders-per-node` browsers) from "receiver" nodes (that will launch at most `receivers-per-node` browsers).

Reply:
Ok, I'll just consider myself here to ask the dumb questions then :) 

> This script is using it to distinguish "sender" nodes (i.e. ones that will launch at most senders-per-node browsers) from "receiver" nodes (that will launch at most receivers-per-node browsers).

What about endpoints that we want to send and receive?

Reply:
"Sender" means send-and-receive; we don't have send-only nodes currently.  (Given that we're just exercising the jitsi-meet UI in this test, is there even a way to configure a node as send-only?)

Reply:
So is it that: we get some configuration from the command line which has some amount of total senders and receivers, and then a setting of the maximum amount of senders per node and the maximum amount of receivers per node, and then the backend grabs a node and then:

1) Checks what it's current configuration is to see if it fits within senders-per-node and receivers-per-node, and then, if necessary, changes the amount of browsers spun up on that node?

or 
2) Just overrides whatever config exists to make this node conform to senders-per-node and receivers-per-node?

Reply:
It's option 2. It doesn't do any checking.

I don't think "see if it fits within" is a good idea; one use case I was concerned about is when you discover that the limits you set for a previous run turned out to be too aggressive, and you need to set them lower.  (This is why I also have the script to monitor the Selenium load average in this patch set.)